# Project: Fraud Detection System
# Data Cleaning and Preprocessing

Table of Contents
1. Import libraries and Load Dataset
2. Data Cleaning
    2.1. Data Inspection
    2.2. Duplciate Values
    2.3. Missing Values
        2.3.1. Missing Values Summary
        2.3.2. Missing Values Handling
    2.4. Leakage Columns
    2.5. Outliers

    Dataset Source: https://www.kaggle.com/datasets/aryan208/financial-transactions-dataset-for-fraud-detection

# I. Import Libraries and Load Dataset

In [ ]:
# Import libraries
import os
import time
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
# Define directory path
BASE_DIR = os.path.dirname(os.path.abspath(os.getcwd()))

# Define path to output folder
OUTPUT_DIR = os.path.join(BASE_DIR, "outputs")
DATA_DIR = os.path.join(BASE_DIR, "data")
os.makedirs(OUTPUT_DIR, exist_ok=True)
os.makedirs(DATA_DIR, exist_ok=True)
FILE_PATH = "transactions.csv"

# Print the base directory
print(f"Base directory: {BASE_DIR}")
print(f"Data directory: {DATA_DIR}")
print(f"Output directory: {OUTPUT_DIR}")

Base directory: c:\Users\Nafis\BuildersLab_projects\Teams\Fraud\Fraud-Detection-System
Data directory: c:\Users\Nafis\BuildersLab_projects\Teams\Fraud\Fraud-Detection-System\data
Output directory: c:\Users\Nafis\BuildersLab_projects\Teams\Fraud\Fraud-Detection-System\outputs


In [ ]:
# Import kagglehub
"""import sys
print(sys.executable)
!{sys.executable} -m pip install kagglehub"""

'import sys\nprint(sys.executable)\n!{sys.executable} -m pip install kagglehub'

In [ ]:
# Import the dataset from Kaggle
import kagglehub
from kagglehub import KaggleDatasetAdapter

# Download latest version
path = kagglehub.dataset_download("aryan208/financial-transactions-dataset-for-fraud-detection")

# Check the path where the dataset files are stored
print("Path to dataset files:", path)

# Check the files that are available in the dataset
print(os.listdir(path))

c:\Users\Nafis\BuildersLab_projects\Teams\Fraud\Fraud-Detection-System\.venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Path to dataset files: C:\Users\Nafis\.cache\kagglehub\datasets\aryan208\financial-transactions-dataset-for-fraud-detection\versions\1
['financial_fraud_detection_dataset.csv']


In [ ]:
# Start timer to measure the time taken to load the dataset
start_time = time.time()

# Load dataset
df = pd.read_csv(os.path.join(path, "financial_fraud_detection_dataset.csv"))

# End timer and print the time taken to load the dataset
end_time = time.time()
print(f"Time taken to load the dataset: {end_time - start_time:.2f} seconds")

Time taken to load the dataset: 26.13 seconds


#II. Data Quality
1. Import libraries and Load Dataset
2. Data Cleaning
    2.1. Data Inspection
    2.2. Duplciate Values
    2.3. Missing Values
        2.3.1. Missing Values Summary
        2.3.2. Missing Values Handling
    2.4. Leakage Columns
    2.5. Outliers

### 1. Data Inspection

In [ ]:
# Check the shape of the dataset
print("Shape of the dataset:", df.shape)

Shape of the dataset: (5000000, 18)


In [ ]:
# Check a sample of the dataset
print("Sample of the dataset:")
df.sample(5)

Sample of the dataset:


,transaction_id,timestamp,sender_account,receiver_account,amount,transaction_type,merchant_category,location,device_used,is_fraud,fraud_type,time_since_last_transaction,spending_deviation_score,velocity_score,geo_anomaly_score,payment_channel,ip_address,device_hash
512381,T612381,2023-07-20T05:14:36.704687,ACC508702,ACC457407,81.08,payment,other,Berlin,web,False,NaN,NaN,0.60,4,0.06,wire_transfer,15.38.211.35,D3835343
1334211,T1434211,2023-11-05T19:37:50.122027,ACC659103,ACC717163,1020.82,deposit,utilities,Dubai,atm,False,NaN,6530.839794,-0.19,20,0.62,ACH,82.164.71.63,D6009574
924666,T1024666,2023-02-03T23:19:48.178048,ACC755652,ACC208679,32.12,transfer,utilities,Berlin,mobile,False,NaN,NaN,1.06,19,0.42,wire_transfer,251.181.34.67,D3321361
4565079,T4665079,2023-11-03T23:21:00.178276,ACC100741,ACC672259,127.56,withdrawal,travel,New York,mobile,False,NaN,5534.504121,-0.60,8,0.99,UPI,38.160.226.179,D9047743
1055801,T1155801,2023-05-11T17:07:00.850265,ACC567455,ACC920618,133.51,deposit,travel,Tokyo,atm,False,NaN,-5118.622466,2.51,17,0.49,ACH,97.35.145.89,D3868503


In [ ]:
# Display all columns
print("Columns in the dataset:")
print(df.columns.tolist())

Columns in the dataset:
['transaction_id', 'timestamp', 'sender_account', 'receiver_account', 'amount', 'transaction_type', 'merchant_category', 'location', 'device_used', 'is_fraud', 'fraud_type', 'time_since_last_transaction', 'spending_deviation_score', 'velocity_score', 'geo_anomaly_score', 'payment_channel', 'ip_address', 'device_hash']


In [ ]:
# Check info
print("Info about the dataset:")
df.info(verbose=True, show_counts=True) # Display all columns and count of non-null values

Info about the dataset:
<class 'pandas.DataFrame'>
RangeIndex: 5000000 entries, 0 to 4999999
Data columns (total 18 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   transaction_id               5000000 non-null  str    
 1   timestamp                    5000000 non-null  str    
 2   sender_account               5000000 non-null  str    
 3   receiver_account             5000000 non-null  str    
 4   amount                       5000000 non-null  float64
 5   transaction_type             5000000 non-null  str    
 6   merchant_category            5000000 non-null  str    
 7   location                     5000000 non-null  str    
 8   device_used                  5000000 non-null  str    
 9   is_fraud                     5000000 non-null  bool   
 10  fraud_type                   179553 non-null   str    
 11  time_since_last_transaction  4103487 non-null  float64
 12  spending_deviation_score     

In [ ]:
# Display summary statistics for numeric columns
df.describe()

,amount,time_since_last_transaction,spending_deviation_score,velocity_score,geo_anomaly_score
count,5.000000e+06,4.103487e+06,5.000000e+06,5.000000e+06,5.000000e+06
mean,3.589343e+02,1.525799e+00,-3.881160e-04,1.050132e+01,5.000293e-01
std,4.699333e+02,3.576569e+03,1.000807e+00,5.766842e+00,2.886349e-01
min,1.000000e-02,-8.777814e+03,-5.260000e+00,1.000000e+00,0.000000e+00
25%,2.657000e+01,-2.562376e+03,-6.800000e-01,5.000000e+00,2.500000e-01
50%,1.386700e+02,8.442747e-01,0.000000e+00,1.100000e+01,5.000000e-01
75%,5.038900e+02,2.568339e+03,6.700000e-01,1.600000e+01,7.500000e-01
max,3.520570e+03,8.757758e+03,5.020000e+00,2.000000e+01,1.000000e+00


In [ ]:
# Print all the values of the 'fraud_type' column
print(f"Unique values in 'fraud_type' column:")
print(df['fraud_type'].unique().tolist())

# Print all the values of the 'is_fraud' column
print("\nUnique values in 'is_fraud' column:")
print(df['is_fraud'].unique().tolist())

Unique values in 'fraud_type' column:
[nan, 'card_not_present']

Unique values in 'is_fraud' column:
[False, True]


In [ ]:
# Count and percentage of values in 'is_fraud' column
fraud_counts = df['is_fraud'].value_counts()
fraud_pcts = df['is_fraud'].value_counts(normalize=True) * 100

fraud_summary = pd.DataFrame({
    "count": fraud_counts,
    "percentage": fraud_pcts.round(2)
})

print("Summary of 'is_fraud' column:")
fraud_summary

Summary of 'is_fraud' column:


,count,percentage
is_fraud,,
False,4820447,96.41
True,179553,3.59


### 2. Duplicate Values

In [ ]:
# Check for duplicate rows
print("Number of duplicate rows:", df.duplicated().sum())

# Check for duplicate values in the "id" column
print("Number of duplicate values in 'transactions_id' column:", df['transaction_id'].duplicated().sum())

Number of duplicate rows: 0
Number of duplicate values in 'transactions_id' column: 0


### 3. Missing Values
#### 3.1. Missing Values Summary

In [ ]:
# Summary of missing values
missing_values = df.isnull().sum()
missing_values_percentage = (missing_values / len(df)) * 100

# Create a DataFrame to display the summary of missing values
missing_values_summary = pd.DataFrame({
    'Missing Values': missing_values,
    'Percentage (%)': missing_values_percentage
}).sort_values(by='Missing Values', ascending=False)

# Add 'column_name' as a column in the DataFrame
missing_values_summary['Column names'] = missing_values_summary.index

# Save to a csv file to outputs folder
missing_values_summary.to_csv(os.path.join(OUTPUT_DIR, "missing_values_summary.csv"))

print("Summary of missing values:")
missing_values_summary

Summary of missing values:


,Missing Values,Percentage (%),Column names
fraud_type,4820447,96.40894,fraud_type
time_since_last_transaction,896513,17.93026,time_since_last_transaction
transaction_id,0,0.00000,transaction_id
timestamp,0,0.00000,timestamp
ip_address,0,0.00000,ip_address
payment_channel,0,0.00000,payment_channel
geo_anomaly_score,0,0.00000,geo_anomaly_score
velocity_score,0,0.00000,velocity_score
spending_deviation_score,0,0.00000,spending_deviation_score
is_fraud,0,0.00000,is_fraud


In [ ]:
# Inspect missingness in fradu_type
# Check if fraud type is missing whenever transaction is_fraud = 1
# Missing fraud_type by is_fraud
missing_summary = df.groupby('is_fraud')['fraud_type'].apply(lambda x: x.isna().sum()).to_frame('missing_count')
missing_summary['missing_pct'] = (missing_summary['missing_count'] / df.groupby('is_fraud').size() * 100).round(2)

missing_summary

,missing_count,missing_pct
is_fraud,,
False,4820447,100.0
True,0,0.0


Comment: fraud_type is missing for 100% of non-fraud rows and present for 100% of fraud rows.

In [ ]:
# Inspect missingness by the sender's total transaction count
person_col = 'sender_account'

# Create a copy of the dataset
df_eda = df.copy()

# Count total transactions for each sender
transaction_counts = df_eda[person_col].value_counts()
df_eda['transaction_count'] = df_eda[person_col].map(transaction_counts)

# Group senders with exactly one transaction versus senders with more than one
df_eda['transaction_group'] = np.where(
    df_eda['transaction_count'] == 1,
    '1 transaction',
    'more than 1 transaction'
 )

# Calculate missing counts and percentages across transactions in each group
missing_time_by_transaction_group = (
    df_eda.groupby('transaction_group', observed=True)['time_since_last_transaction']
      .agg(
          missing_count=lambda values: values.isna().sum(),
          total_transactions='size'
      )
 )
missing_time_by_transaction_group['missing_pct'] = (
    missing_time_by_transaction_group['missing_count']
    / missing_time_by_transaction_group['total_transactions']
    * 100
 ).round(2)

missing_time_by_transaction_group

,missing_count,total_transactions,missing_pct
transaction_group,,,
1 transaction,19210,19210,100.00
more than 1 transaction,877303,4980790,17.61


Comments:
- 100% of senders with exactly one transaction have a missing value in `time_since_last_transaction`. This is expected because there is no previous transaction from which to calculate the time since the last transaction.
- 17.61% of transaction rows belonging to senders with more than one transaction have a missing value in `time_since_last_transaction`. Because these senders have transaction history, we will inspect whether the missing values occur on their first transaction or on later transactions.

In [ ]:
# For repeat senders, check whether missingness occurs on the first or a later transaction
repeat_df = df_eda[df_eda['transaction_count'] > 1].copy()
repeat_df['timestamp'] = pd.to_datetime(
    repeat_df['timestamp'],
    format='mixed'
 )
repeat_df = repeat_df.sort_values([person_col, 'timestamp', 'transaction_id'])

# Assign each repeat sender a chronological transaction number
repeat_df['transaction_number'] = (
    repeat_df.groupby(person_col).cumcount() + 1
 )
repeat_df['transaction_position'] = np.where(
    repeat_df['transaction_number'] == 1,
    'first transaction',
    'later transaction'
 )

# Calculate missingness by transaction position among repeat senders
missing_time_by_position_repeat = (
    repeat_df.groupby('transaction_position', observed=True)['time_since_last_transaction']
      .agg(
          missing_count=lambda values: values.isna().sum(),
          total_transactions='size'
      )
 )
missing_time_by_position_repeat['missing_pct'] = (
    missing_time_by_position_repeat['missing_count']
    / missing_time_by_position_repeat['total_transactions']
    * 100
 ).round(2)

missing_time_by_position_repeat

,missing_count,total_transactions,missing_pct
transaction_position,,,
first transaction,185427,877303,21.14
later transaction,691876,4103487,16.86


In [ ]:
# Explore whether missingness is related to observed variables
analysis_df = df_eda.copy()
analysis_df['time_since_last_transaction_missing'] = (
    analysis_df['time_since_last_transaction'].isna()
 )
analysis_df['amount_decile'] = pd.qcut(
    analysis_df['amount'],
    q=10,
    duplicates='drop'
 )
analysis_df['transaction_hour'] = pd.to_datetime(
    analysis_df['timestamp'],
    format='mixed'
 ).dt.hour

observed_features = [
    'transaction_count',
    'transaction_group',
    'amount_decile',
    'transaction_type',
    'merchant_category',
    'location',
    'device_used',
    'transaction_hour',
    'is_fraud'
 ]

print(
    'Overall missing percentage:',
    f"{analysis_df['time_since_last_transaction_missing'].mean() * 100:.2f}%"
 )

for feature in observed_features:
    feature_summary = (
        analysis_df.groupby(feature, observed=True)['time_since_last_transaction_missing']
          .agg(
              missing_count='sum',
              total_rows='size'
          )
     )
    feature_summary['missing_pct'] = (
        feature_summary['missing_count']
        / feature_summary['total_rows']
        * 100
     ).round(2)
    print(f'\nMissingness by {feature}:')
    display(feature_summary)

Overall missing percentage: 17.93%

Missingness by transaction_count:


,missing_count,total_rows,missing_pct
transaction_count,,,
1,19210,19210,100.00
2,53751,107502,50.00
3,99548,298644,33.33
4,138234,552936,25.00
5,153544,767720,20.00
6,142359,854154,16.67
7,112483,787381,14.29
8,77433,619464,12.50
9,48616,437544,11.11



Missingness by transaction_group:


,missing_count,total_rows,missing_pct
transaction_group,,,
1 transaction,19210,19210,100.00
more than 1 transaction,877303,4980790,17.61



Missingness by amount_decile:


,missing_count,total_rows,missing_pct
amount_decile,,,
"(0.009000000000000001, 11.05]",89960,500093,17.99
"(11.05, 20.4]",89690,500011,17.94
"(20.4, 34.93]",89334,500024,17.87
"(34.93, 67.13]",90023,499905,18.01
"(67.13, 138.67]",89465,500007,17.89
"(138.67, 254.58]",89729,499997,17.95
"(254.58, 405.11]",89470,499988,17.89
"(405.11, 649.55]",89545,499987,17.91
"(649.55, 1126.5]",89354,499995,17.87



Missingness by transaction_type:


,missing_count,total_rows,missing_pct
transaction_type,,,
deposit,224237,1250593,17.93
payment,223947,1250438,17.91
transfer,224383,1250334,17.95
withdrawal,223946,1248635,17.94



Missingness by merchant_category:


,missing_count,total_rows,missing_pct
merchant_category,,,
entertainment,112226,625332,17.95
grocery,112252,624954,17.96
online,111421,623581,17.87
other,111431,624589,17.84
restaurant,111879,625483,17.89
retail,113112,626319,18.06
travel,112051,625656,17.91
utilities,112141,624086,17.97



Missingness by location:


,missing_count,total_rows,missing_pct
location,,,
Berlin,112205,625289,17.94
Dubai,112161,624320,17.97
London,111636,624256,17.88
New York,112157,625354,17.93
Singapore,112515,625313,17.99
Sydney,111469,625125,17.83
Tokyo,112572,625994,17.98
Toronto,111798,624349,17.91



Missingness by device_used:


,missing_count,total_rows,missing_pct
device_used,,,
atm,224326,1249640,17.95
mobile,223934,1251131,17.90
pos,224296,1249158,17.96
web,223957,1250071,17.92



Missingness by transaction_hour:


,missing_count,total_rows,missing_pct
transaction_hour,,,
0,37381,208141,17.96
1,37374,207952,17.97
2,37464,208680,17.95
3,37482,208256,18.00
4,37156,208532,17.82
5,37738,209498,18.01
6,37250,208093,17.90
7,37255,208114,17.90
8,37607,208967,18.00



Missingness by is_fraud:


,missing_count,total_rows,missing_pct
is_fraud,,,
False,896513,4820447,18.6
True,0,179553,0.0


Comments:
- The missingness is not completely random (MCAR): missingness varies systematically with the sender's transaction count, with exactly one missing value per sender.
- Missingness is strongly associated with the target: all fraudulent transactions have a non-missing `time_since_last_transaction`, while all missing values occur on non-fraudulent transactions. This suggests a synthetic data-generation rule or target leakage and requires further investigation before modeling.

### Outliers

In [ ]:
# Detect potential outliers in numeric transaction features using the IQR rule
numeric_columns = [
    'amount',
    'time_since_last_transaction',
    'spending_deviation_score',
    'velocity_score',
    'geo_anomaly_score'
 ]

outlier_summary = []

for column in numeric_columns:
    values = df[column].dropna()
    q1 = values.quantile(0.25)
    q3 = values.quantile(0.75)
    iqr = q3 - q1
    lower_bound = q1 - 1.5 * iqr
    upper_bound = q3 + 1.5 * iqr
    outlier_mask = (df[column] < lower_bound) | (df[column] > upper_bound)

    outlier_summary.append({
        'column': column,
        'q1': q1,
        'q3': q3,
        'lower_bound': lower_bound,
        'upper_bound': upper_bound,
        'outlier_count': outlier_mask.sum(),
        'outlier_pct': outlier_mask.mean() * 100
     })

outlier_summary = (
    pd.DataFrame(outlier_summary)
      .sort_values('outlier_count', ascending=False)
 )
outlier_summary['outlier_pct'] = outlier_summary['outlier_pct'].round(2)

outlier_summary

,column,q1,q3,lower_bound,upper_bound,outlier_count,outlier_pct
0,amount,26.570000,503.890000,-689.410000,1219.87000,412449,8.25
2,spending_deviation_score,-0.680000,0.670000,-2.705000,2.69500,35075,0.70
1,time_since_last_transaction,-2562.376056,2568.338671,-10258.448146,10264.41076,0,0.00
3,velocity_score,5.000000,16.000000,-11.500000,32.50000,0,0.00
4,geo_anomaly_score,0.250000,0.750000,-0.500000,1.50000,0,0.00


Comments:

In [ ]:
# Check domain-validity issues separately from statistical outliers
domain_checks = pd.DataFrame({
    'check': [
        'Negative time_since_last_transaction',
        'Non-positive amount',
        'Minimum amount',
        'Maximum amount',
        'Minimum time_since_last_transaction',
        'Maximum time_since_last_transaction'
     ],
    'value': [
        (df['time_since_last_transaction'] < 0).sum(),
        (df['amount'] <= 0).sum(),
        df['amount'].min(),
        df['amount'].max(),
        df['time_since_last_transaction'].min(),
        df['time_since_last_transaction'].max()
     ]
 })

domain_checks

,check,value
0,Negative time_since_last_transaction,2.051331e+06
1,Non-positive amount,0.000000e+00
2,Minimum amount,1.000000e-02
3,Maximum amount,3.520570e+03
4,Minimum time_since_last_transaction,-8.777814e+03
5,Maximum time_since_last_transaction,8.757758e+03


Comment
- `amount` has 412,449 high values, which is 8.25% of the data. These may be real large transactions, so we should not remove them automatically.
- `spending_deviation_score` has 35,075 possible outliers, which is 0.70% of the data.
- `velocity_score` and `geo_anomaly_score` have no values flagged by the IQR method.
- `time_since_last_transaction` has 2,051,331 negative values. This is not valid for a time-since feature, so the column should be checked or recreated from the timestamps.
- We will not remove outliers yet. First, we will decide whether each value is unusual or actually incorrect.

In [ ]:
# Create a working dataframe without the untrusted time-since feature
df_clean = df.drop(columns=['time_since_last_transaction']).copy()

print('Removed column: time_since_last_transaction')
print(f'Cleaned dataframe shape: {df_clean.shape}')

Removed column: time_since_last_transaction
Cleaned dataframe shape: (5000000, 17)


## Validate `time_since_last_transaction`

The original value is compared with a value recalculated from timestamps after sorting each sender chronologically. The first transaction for each sender has no previous transaction, so its recalculated value is missing.

In [ ]:
# Recalculate time since the sender's previous transaction
time_check = df[[
    'transaction_id',
    'sender_account',
    'timestamp',
    'time_since_last_transaction'
]].copy()

time_check['timestamp'] = pd.to_datetime(
    time_check['timestamp'],
    format='mixed',
    errors='coerce'
)
time_check['original_time_since_last_transaction'] = pd.to_numeric(
    time_check['time_since_last_transaction'],
    errors='coerce'
)

# Sort transactions in chronological order for each sender
time_check = (
    time_check
    .sort_values(['sender_account', 'timestamp', 'transaction_id'])
    .reset_index(drop=True)
)

time_check['recalculated_time_since_last_transaction'] = (
    time_check.groupby('sender_account')['timestamp']
    .diff()
    .dt.total_seconds()
)

# Compare values
time_check['matches'] = (
    np.isclose(
        time_check['original_time_since_last_transaction'].fillna(-1),
        time_check['recalculated_time_since_last_transaction'].fillna(-1),
        rtol=0,
        atol=1
    )
)

comparison_summary = pd.DataFrame({
    'metric': [
        'total rows',
        'matching rows',
        'different rows',
        'original negative values',
        'recalculated negative values',
        'original missing values',
        'recalculated missing values'
    ],
    'value': [
        len(time_check),
        time_check['matches'].sum(),
        (~time_check['matches']).sum(),
        (time_check['original_time_since_last_transaction'] < 0).sum(),
        (time_check['recalculated_time_since_last_transaction'] < 0).sum(),
        time_check['original_time_since_last_transaction'].isna().sum(),
        time_check['recalculated_time_since_last_transaction'].isna().sum()
    ]
})

display(comparison_summary)

# Inspect the first mismatches
mismatches = time_check.loc[
    ~time_check['matches'],
    [
        'transaction_id',
        'sender_account',
        'timestamp',
        'original_time_since_last_transaction',
        'recalculated_time_since_last_transaction'
    ]
].copy()
mismatches['difference_seconds'] = (
    mismatches['original_time_since_last_transaction']
    - mismatches['recalculated_time_since_last_transaction']
)

display(mismatches.head(20))
mismatches.to_csv(
    os.path.join(OUTPUT_DIR, 'time_since_transaction_mismatches.csv'),
    index=False
)

,metric,value
0,total rows,5000000
1,matching rows,204858
2,different rows,4795142
3,original negative values,2051331
4,recalculated negative values,0
5,original missing values,896513
6,recalculated missing values,896513


,transaction_id,sender_account,timestamp,original_time_since_last_transaction,recalculated_time_since_last_transaction,difference_seconds
2,T991203,ACC100001,2023-12-27 21:20:18.603045,2731.846303,9.834647e+06,-9.831915e+06
4,T631422,ACC100002,2023-08-22 08:37:49.428281,2323.179069,8.363445e+06,-8.361121e+06
5,T5062367,ACC100002,2023-09-30 15:45:51.319369,943.133859,3.395282e+06,-3.394339e+06
6,T1170676,ACC100003,2023-04-13 14:38:00.520010,-3218.619256,NaN,NaN
7,T1003981,ACC100003,2023-04-30 03:53:31.626724,NaN,1.430131e+06,NaN
8,T3303332,ACC100003,2023-05-21 00:57:48.160305,-5164.552279,1.803857e+06,-1.809021e+06
9,T1149234,ACC100003,2023-08-25 17:15:09.842970,-211.224876,8.353042e+06,-8.353253e+06
10,T1018914,ACC100003,2023-09-03 12:28:39.396761,3032.585492,7.604096e+05,-7.573770e+05
11,T4836104,ACC100003,2023-12-04 04:56:39.076104,4731.980810,7.921680e+06,-7.916948e+06
12,T1997184,ACC100003,2023-12-22 05:30:56.365279,6062.882179,1.557257e+06,-1.551194e+06


## Data Modeling

Time-based split: The raw transactions are split chronologically before feature engineering. Historical features are calculated using only the current transaction and transactions occurring earlier in time.

In [ ]:
# Split raw transactions before calculating any history-based features
raw_model_data = df.copy()
raw_model_data['timestamp'] = pd.to_datetime(
    raw_model_data['timestamp'], format='mixed', errors='coerce'
)
raw_model_data['is_fraud'] = pd.to_numeric(
    raw_model_data['is_fraud'], errors='coerce'
)
raw_model_data = (
    raw_model_data.dropna(subset=['timestamp', 'is_fraud'])
    .sort_values(['timestamp', 'transaction_id'])
    .reset_index(drop=True)
)
raw_model_data['is_fraud'] = raw_model_data['is_fraud'].astype(int)

n_rows = len(raw_model_data)
train_end = int(n_rows * 0.70)
validation_end = int(n_rows * 0.85)

train_raw = raw_model_data.iloc[:train_end].copy()
validation_raw = raw_model_data.iloc[train_end:validation_end].copy()
test_raw = raw_model_data.iloc[validation_end:].copy()

print(train_raw.shape, validation_raw.shape, test_raw.shape)

(3500000, 18) (750000, 18) (750000, 18)


In [ ]:
def add_prior_features(frame):
    frame = frame.copy()
    frame = frame.sort_values(['timestamp', 'transaction_id']).reset_index(drop=True)

    # Calendar features use only the transaction timestamp.
    frame['hour'] = frame['timestamp'].dt.hour
    frame['day_of_week'] = frame['timestamp'].dt.dayofweek
    frame['month'] = frame['timestamp'].dt.month
    frame['is_weekend'] = frame['day_of_week'].isin([5, 6]).astype(int)
    frame['is_night'] = frame['hour'].isin([0, 1, 2, 3, 4, 5]).astype(int)

    # Prior sender statistics. Current amount is excluded from the history.
    sender = frame.groupby('sender_account', sort=False, dropna=False)
    frame['sender_prior_count'] = sender.cumcount()
    frame['time_since_sender_transaction'] = (
        sender['timestamp'].diff().dt.total_seconds()
    )
    prior_sum = sender['amount'].cumsum() - frame['amount']
    prior_squared_sum = (
        frame['amount'].pow(2).groupby(
            frame['sender_account'], sort=False, dropna=False
        ).cumsum() - frame['amount'].pow(2)
    )
    prior_count = frame['sender_prior_count'].replace(0, np.nan)
    frame['sender_prior_mean'] = prior_sum / prior_count
    frame['sender_prior_variance'] = (
        prior_squared_sum / prior_count - frame['sender_prior_mean'].pow(2)
    ).clip(lower=0)
    frame['sender_prior_std'] = np.sqrt(frame['sender_prior_variance'])
    frame['amount_vs_sender_prior_mean'] = (
        frame['amount'] / frame['sender_prior_mean']
    ).replace([np.inf, -np.inf], np.nan)

    # Explicit "no history yet" flag, since NaN in the columns above means no prior transaction, the imputer needs this
    # flag to tell those rows apart from a real zero-valued observation.
    frame['is_first_sender_transaction'] = (
        frame['sender_prior_count'] == 0
    ).astype(int)

    # Prior receiver and sender-receiver relationship statistics.
    receiver = frame.groupby('receiver_account', sort=False, dropna=False)
    frame['receiver_prior_count'] = receiver.cumcount()
    pair = frame.groupby(
        ['sender_account', 'receiver_account'], sort=False, dropna=False
    )
    frame['sender_receiver_prior_count'] = pair.cumcount()
    frame['time_since_sender_receiver_transaction'] = (
        pair['timestamp'].diff().dt.total_seconds()
    )
    frame['is_first_sender_receiver_pair'] = (
        frame['sender_receiver_prior_count'] == 0
    ).astype(int)

    # Prior device and IP counts.
    for column, prefix in [('device_hash', 'device'), ('ip_address', 'ip')]:
        frame[f'{prefix}_prior_count'] = frame.groupby(
            column, sort=False, dropna=False
        ).cumcount()
        frame[f'is_new_{prefix}'] = (
            frame[f'{prefix}_prior_count'] == 0
        ).astype(int)

    frame['log_amount'] = np.log1p(frame['amount'].clip(lower=0))
    frame['amount_round_number_flag'] = (frame['amount'] % 100 == 0).astype(int)
    frame['rapid_transaction_flag'] = (
        frame['time_since_sender_transaction'] < 60
    ).fillna(False).astype(int)
    return frame

In [ ]:
# Calculate features after splitting
feature_input = pd.concat([
    train_raw.assign(_split='train'),
    validation_raw.assign(_split='validation'),
    test_raw.assign(_split='test')
], ignore_index=True)
featured_data = add_prior_features(feature_input)

train_features = featured_data[featured_data['_split'] == 'train'].copy()
validation_features = featured_data[featured_data['_split'] == 'validation'].copy()
test_features = featured_data[featured_data['_split'] == 'test'].copy()

NON_MODEL_COLUMNS = [
    'is_fraud', '_split', 'timestamp', 'transaction_id',
    'sender_account', 'receiver_account', 'device_hash',
    'ip_address', 'fraud_type', 'time_since_last_transaction',
    'transaction_count', 'transaction_group'
]
feature_columns = [
    c for c in train_features.columns if c not in NON_MODEL_COLUMNS
]

X_train = train_features[feature_columns].copy()
y_train = train_features['is_fraud'].copy()
X_validation = validation_features[feature_columns].copy()
y_validation = validation_features['is_fraud'].copy()
X_test = test_features[feature_columns].copy()
y_test = test_features['is_fraud'].copy()

print(f'Features: {len(feature_columns)}')
print('Train fraud rate:', y_train.mean())
print('Validation fraud rate:', y_validation.mean())
print('Test fraud rate:', y_test.mean())

Features: 32
Train fraud rate: 0.035843714285714284
Validation fraud rate: 0.03602533333333333
Test fraud rate: 0.036108


In [ ]:
# Check for missing values introduced by feature engineering.
train_missing = X_train.isna().sum()
train_missing = train_missing[train_missing > 0].sort_values(ascending=False)
train_missing_pct = (train_missing / len(X_train) * 100).round(2)

missing_after_features = pd.DataFrame({
    "missing_count": train_missing,
    "missing_pct": train_missing_pct
})
print("Columns with missing values in X_train after feature engineering:")
display(missing_after_features)


Columns with missing values in X_train after feature engineering:


,missing_count,missing_pct
time_since_sender_receiver_transaction,3499992,100.00
time_since_sender_transaction,881705,25.19
sender_prior_mean,881705,25.19
sender_prior_variance,881705,25.19
sender_prior_std,881705,25.19
amount_vs_sender_prior_mean,881705,25.19


## Class-Weighted Logistic Regression Baseline

This baseline uses `class_weight='balanced'`

In [ ]:
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import (
    average_precision_score, classification_report,
    confusion_matrix, precision_recall_curve, roc_auc_score
)

numeric_features = X_train.select_dtypes(include='number').columns.tolist()
categorical_features = X_train.select_dtypes(
    include=['object', 'category', 'string']
).columns.tolist()

preprocessor = ColumnTransformer([
    ('numeric', Pipeline([
        ('imputer', SimpleImputer(strategy='median')),
        ('scaler', StandardScaler())
    ]), numeric_features),
    ('categorical', Pipeline([
        ('imputer', SimpleImputer(strategy='most_frequent')),
        ('encoder', OneHotEncoder(handle_unknown='ignore', sparse_output=True))
    ]), categorical_features)
])

baseline_model = Pipeline([
    ('preprocessor', preprocessor),
    ('classifier', LogisticRegression(
        class_weight='balanced', max_iter=1000,
        solver='liblinear', random_state=42
    ))
])

baseline_model.fit(X_train, y_train)
validation_probability = baseline_model.predict_proba(X_validation)[:, 1]

print('Validation ROC-AUC:', roc_auc_score(y_validation, validation_probability))
print('Validation PR-AUC:', average_precision_score(y_validation, validation_probability))

Validation ROC-AUC: 0.5234747718657466
Validation PR-AUC: 0.038322841574983495


In [ ]:
# Tune the classification threshold on validation data using F1
precision, recall, thresholds = precision_recall_curve(
    y_validation, validation_probability
)
f1_scores = 2 * precision[:-1] * recall[:-1] / (
    precision[:-1] + recall[:-1] + 1e-12
)
best_index = np.argmax(f1_scores)
best_threshold = thresholds[best_index]
validation_prediction = (
    validation_probability >= best_threshold
).astype(int)

print(f'Best validation threshold: {best_threshold:.4f}')
print(classification_report(
    y_validation, validation_prediction,
    target_names=['non-fraud', 'fraud'], zero_division=0
))
print('Confusion matrix:')
print(confusion_matrix(y_validation, validation_prediction))

Best validation threshold: 0.4880
              precision    recall  f1-score   support

   non-fraud       0.97      0.30      0.46    722981
       fraud       0.04      0.73      0.07     27019

    accuracy                           0.32    750000
   macro avg       0.50      0.52      0.27    750000
weighted avg       0.93      0.32      0.45    750000

Confusion matrix:
[[219876 503105]
 [  7238  19781]]


In [ ]:
# Evaluate the selected threshold once on the untouched test set
test_probability = baseline_model.predict_proba(X_test)[:, 1]
test_prediction = (test_probability >= best_threshold).astype(int)

test_report = classification_report(
    y_test, test_prediction, output_dict=True, zero_division=0
)
test_cm = confusion_matrix(y_test, test_prediction)
tn, fp, fn, tp = test_cm.ravel()

test_precision = test_report['1']['precision']
test_recall = test_report['1']['recall']
test_f1 = test_report['1']['f1-score']
test_f2 = (5 * test_precision * test_recall) / (
    4 * test_precision + test_recall + 1e-12
)
test_specificity = tn / (tn + fp + 1e-12)
test_fpr = fp / (fp + tn + 1e-12)
test_fnr = fn / (fn + tp + 1e-12)

print('Test ROC-AUC:', roc_auc_score(y_test, test_probability))
print('Test PR-AUC:', average_precision_score(y_test, test_probability))
print(f'Test threshold: {best_threshold:.4f}')
print(classification_report(
    y_test, test_prediction,
    target_names=['non-fraud', 'fraud'], zero_division=0
))
print('Test confusion matrix:')
print(test_cm)

Test ROC-AUC: 0.5223939814215609
Test PR-AUC: 0.038273149312674476
Test threshold: 0.4880
              precision    recall  f1-score   support

   non-fraud       0.97      0.37      0.53    722919
       fraud       0.04      0.66      0.07     27081

    accuracy                           0.38    750000
   macro avg       0.50      0.52      0.30    750000
weighted avg       0.93      0.38      0.52    750000

Test confusion matrix:
[[265424 457495]
 [  9096  17985]]


## Metrics Comparison Table

The table includes ranking metrics from predicted probabilities and classification metrics from the selected validation threshold. The test set remains untouched until final evaluation.

In [ ]:
# Save validation and test performance in one comparison table
validation_report = classification_report(
    y_validation, validation_prediction, output_dict=True, zero_division=0
)
validation_cm = confusion_matrix(y_validation, validation_prediction)
val_tn, val_fp, val_fn, val_tp = validation_cm.ravel()
val_precision = validation_report['1']['precision']
val_recall = validation_report['1']['recall']
val_f1 = validation_report['1']['f1-score']
val_specificity = val_tn / (val_tn + val_fp + 1e-12)

metrics_table = pd.DataFrame([
    {
        'model': 'Logistic Regression',
        'strategy': 'class_weight=balanced',
        'dataset': 'validation',
        'threshold': best_threshold,
        'precision': val_precision,
        'recall': val_recall,
        'f1': val_f1,
        'f2': (5 * val_precision * val_recall) / (4 * val_precision + val_recall + 1e-12),
        'pr_auc': average_precision_score(y_validation, validation_probability),
        'roc_auc': roc_auc_score(y_validation, validation_probability),
        'specificity': val_specificity,
        'false_positive_rate': val_fp / (val_fp + val_tn + 1e-12),
        'false_negative_rate': val_fn / (val_fn + val_tp + 1e-12)
    },
    {
        'model': 'Logistic Regression',
        'strategy': 'class_weight=balanced',
        'dataset': 'test',
        'threshold': best_threshold,
        'precision': test_precision,
        'recall': test_recall,
        'f1': test_f1,
        'f2': test_f2,
        'pr_auc': average_precision_score(y_test, test_probability),
        'roc_auc': roc_auc_score(y_test, test_probability),
        'specificity': test_specificity,
        'false_positive_rate': test_fpr,
        'false_negative_rate': test_fnr
    }
])

metrics_table = metrics_table.round(4)
display(metrics_table)
metrics_table.to_csv(
    os.path.join(OUTPUT_DIR, 'model_metrics_comparison.csv'),
    index=False
)
print('Metrics saved to:', os.path.join(OUTPUT_DIR, 'model_metrics_comparison.csv'))

,model,strategy,dataset,threshold,precision,recall,f1,f2,pr_auc,roc_auc,specificity,false_positive_rate,false_negative_rate
0,Logistic Regression,class_weight=balanced,validation,0.488,0.0378,0.7321,0.0719,0.1568,0.0383,0.5235,0.3041,0.6959,0.2679
1,Logistic Regression,class_weight=balanced,test,0.488,0.0378,0.6641,0.0716,0.1540,0.0383,0.5224,0.3672,0.6328,0.3359


Metrics saved to: c:\Users\Nafis\BuildersLab_projects\Teams\Fraud\Fraud-Detection-System\outputs\model_metrics_comparison.csv


In [ ]:
# Plot and save probability-ranking curves for model comparison
from sklearn.metrics import RocCurveDisplay, PrecisionRecallDisplay

fig, axes = plt.subplots(1, 2, figsize=(14, 5))
PrecisionRecallDisplay.from_predictions(y_test, test_probability, ax=axes[0])
axes[0].set_title("Test Precision-Recall Curve")
RocCurveDisplay.from_predictions(y_test, test_probability, ax=axes[1])
axes[1].set_title("Test ROC Curve")
plt.tight_layout()
curves_path = os.path.join(OUTPUT_DIR, "baseline_evaluation_curves.png")
plt.savefig(curves_path, dpi=150, bbox_inches="tight")
plt.show()
print("Evaluation curves saved to:", curves_path)

<Figure size 1400x500 with 2 Axes>

Evaluation curves saved to: c:\Users\Nafis\BuildersLab_projects\Teams\Fraud\Fraud-Detection-System\outputs\baseline_evaluation_curves.png


In [ ]:
# Save the baseline model, threshold, feature list, and metrics
import joblib
import json

model_path = os.path.join(OUTPUT_DIR, "fraud_baseline_model.joblib")
threshold_path = os.path.join(OUTPUT_DIR, "fraud_baseline_threshold.json")
features_path = os.path.join(OUTPUT_DIR, "fraud_baseline_features.json")

joblib.dump(baseline_model, model_path)
with open(threshold_path, "w", encoding="utf-8") as file:
    json.dump({"threshold": float(best_threshold)}, file, indent=2)
with open(features_path, "w", encoding="utf-8") as file:
    json.dump(feature_columns, file, indent=2)

print("Model saved to:", model_path)
print("Threshold saved to:", threshold_path)
print("Feature list saved to:", features_path)

Model saved to: c:\Users\Nafis\BuildersLab_projects\Teams\Fraud\Fraud-Detection-System\outputs\fraud_baseline_model.joblib
Threshold saved to: c:\Users\Nafis\BuildersLab_projects\Teams\Fraud\Fraud-Detection-System\outputs\fraud_baseline_threshold.json
Feature list saved to: c:\Users\Nafis\BuildersLab_projects\Teams\Fraud\Fraud-Detection-System\outputs\fraud_baseline_features.json


## Imbalance and Model Comparison

All resampling is applied only to the training matrix. Validation and test data retain their natural fraud rate. The test set is evaluated only after selecting the approach and threshold using validation data.

In [ ]:
# Fit preprocessing once using training data only
X_train_encoded = preprocessor.fit_transform(X_train)
X_validation_encoded = preprocessor.transform(X_validation)
X_test_encoded = preprocessor.transform(X_test)

print("Encoded train shape:", X_train_encoded.shape)
print("Encoded validation shape:", X_validation_encoded.shape)
print("Encoded test shape:", X_test_encoded.shape)

Encoded train shape: (3500000, 55)
Encoded validation shape: (750000, 55)
Encoded test shape: (750000, 55)


In [ ]:
from imblearn.over_sampling import RandomOverSampler, SMOTE
from sklearn.base import clone
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    average_precision_score, balanced_accuracy_score, classification_report,
    confusion_matrix, f1_score, precision_score, recall_score, roc_auc_score
)
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier

def select_f1_threshold(labels, probabilities):
    precision_values, recall_values, threshold_values = precision_recall_curve(
        labels, probabilities
    )
    f1_values = (
        2 * precision_values[:-1] * recall_values[:-1]
        / (precision_values[:-1] + recall_values[:-1] + 1e-12)
    )
    index = int(np.nanargmax(f1_values))
    return float(threshold_values[index])

def calculate_metrics(name, strategy, labels, probabilities, threshold):
    predictions = (probabilities >= threshold).astype(int)
    tn_value, fp_value, fn_value, tp_value = confusion_matrix(
        labels, predictions, labels=[0, 1]
    ).ravel()
    precision_value = precision_score(labels, predictions, zero_division=0)
    recall_value = recall_score(labels, predictions, zero_division=0)
    return {
        "model": name,
        "strategy": strategy,
        "threshold": threshold,
        "precision": precision_value,
        "recall": recall_value,
        "f1": f1_score(labels, predictions, zero_division=0),
        "f2": (5 * precision_value * recall_value) / (4 * precision_value + recall_value + 1e-12),
        "pr_auc": average_precision_score(labels, probabilities),
        "roc_auc": roc_auc_score(labels, probabilities),
        "balanced_accuracy": balanced_accuracy_score(labels, predictions),
        "specificity": tn_value / (tn_value + fp_value + 1e-12),
        "false_positive_rate": fp_value / (fp_value + tn_value + 1e-12),
        "false_negative_rate": fn_value / (fn_value + tp_value + 1e-12)
    }

def fit_and_score(model, name, strategy, train_matrix, train_labels):
    model.fit(train_matrix, train_labels)
    validation_scores = model.predict_proba(X_validation_encoded)[:, 1]
    threshold = select_f1_threshold(y_validation, validation_scores)
    metrics = calculate_metrics(
        name, strategy, y_validation, validation_scores, threshold
    )
    return model, metrics, validation_scores

In [ ]:
# Random oversampling: training data only; validation remains untouched
random_sampler = RandomOverSampler(sampling_strategy=0.25, random_state=42)
X_train_ros, y_train_ros = random_sampler.fit_resample(
    X_train_encoded, y_train
)
ros_model = LogisticRegression(
    max_iter=1000, solver="liblinear", random_state=42
)
ros_model, ros_metrics, ros_validation_probability = fit_and_score(
    ros_model, "Logistic Regression", "random_oversampling_0.25",
    X_train_ros, y_train_ros
)
print("Original training rows:", X_train_encoded.shape[0])
print("Resampled training rows:", X_train_ros.shape[0])
print(ros_metrics)

TypeError: check_array() got an unexpected keyword argument 'force_all_finite'

In [ ]:
# SMOTE: training data only; using a moderate minority ratio to control memory use
# Note: this is a diagnostic comparison after one-hot encoding. For a production
# experiment, use SMOTENC before one-hot encoding so categories are not interpolated.
smote_sampler = SMOTE(sampling_strategy=0.25, random_state=42, k_neighbors=5)
X_train_smote, y_train_smote = smote_sampler.fit_resample(
    X_train_encoded, y_train
)
smote_model = LogisticRegression(
    max_iter=1000, solver="liblinear", random_state=42
)
smote_model, smote_metrics, smote_validation_probability = fit_and_score(
    smote_model, "Logistic Regression", "smote_0.25",
    X_train_smote, y_train_smote
)
print("SMOTE training rows:", X_train_smote.shape[0])
print(smote_metrics)

In [ ]:
# Train tree-based models with class weighting or scale_pos_weight
from catboost import CatBoostClassifier

negative_count = int((y_train == 0).sum())
positive_count = int((y_train == 1).sum())
fraud_weight = negative_count / max(positive_count, 1)

tree_models = [
    (
        "Random Forest",
        "class_weight=balanced",
        RandomForestClassifier(
            n_estimators=100, max_depth=12, min_samples_leaf=5,
            class_weight="balanced", n_jobs=-1, random_state=42
        )
    ),
    (
        "XGBoost",
        "scale_pos_weight",
        XGBClassifier(
            n_estimators=150, max_depth=6, learning_rate=0.08,
            subsample=0.8, colsample_bytree=0.8,
            scale_pos_weight=fraud_weight, objective="binary:logistic",
            eval_metric="aucpr", n_jobs=-1, random_state=42
        )
    ),
    (
        "LightGBM",
        "scale_pos_weight",
        LGBMClassifier(
            n_estimators=150, learning_rate=0.08, num_leaves=31,
            subsample=0.8, colsample_bytree=0.8,
            scale_pos_weight=fraud_weight, verbosity=-1,
            n_jobs=-1, random_state=42
        )
    ),
    (
        "CatBoost",
        "class_weights",
        CatBoostClassifier(
            iterations=150, depth=8, learning_rate=0.08,
            class_weights=[1.0, fraud_weight],
            loss_function="Logloss", eval_metric="AUC:PR",
            verbose=False, random_seed=42, thread_count=-1
        )
    )
]

# Store classifiers that all expect the already-encoded feature matrix.
baseline_classifier = baseline_model.named_steps["classifier"]
model_runs = [
    ("Logistic Regression", "class_weight=balanced", baseline_classifier, validation_probability),
    ("Logistic Regression", "random_oversampling_0.25", ros_model, ros_validation_probability),
    ("Logistic Regression", "smote_0.25", smote_model, smote_validation_probability)
]
tree_validation_scores = {}

for name, strategy, model in tree_models:
    fitted_model, metrics, scores = fit_and_score(
        model, name, strategy, X_train_encoded, y_train
    )
    model_runs.append((name, strategy, fitted_model, scores))
    tree_validation_scores[name] = scores
    print(name, metrics)

In [ ]:
# Build the validation comparison table and choose the candidate by PR-AUC, then F1
validation_rows = []
for name, strategy, model, scores in model_runs:
    threshold = select_f1_threshold(y_validation, scores)
    validation_rows.append(
        calculate_metrics(name, strategy, y_validation, scores, threshold)
    )

validation_comparison = (
    pd.DataFrame(validation_rows)
    .sort_values(["pr_auc", "f1"], ascending=False)
    .reset_index(drop=True)
)
display(validation_comparison.round(4))
validation_comparison.to_csv(
    os.path.join(OUTPUT_DIR, "model_validation_comparison.csv"),
    index=False
)

selected_name = validation_comparison.loc[0, "model"]
selected_strategy = validation_comparison.loc[0, "strategy"]
selected_threshold = float(validation_comparison.loc[0, "threshold"])
selected_model = next(
    model for name, strategy, model, scores in model_runs
    if name == selected_name and strategy == selected_strategy
)
print("Selected candidate:", selected_name, selected_strategy)

In [ ]:
# Evaluate the selected candidate once on the untouched test set
selected_test_probability = selected_model.predict_proba(X_test_encoded)[:, 1]
test_metrics = calculate_metrics(
    selected_name, selected_strategy, y_test,
    selected_test_probability, selected_threshold
)
test_comparison = pd.DataFrame([{**test_metrics, "dataset": "test"}])
display(test_comparison.round(4))

# Append the final test result to the comparison artifact
metrics_table = pd.concat([
    validation_comparison.assign(dataset="validation"),
    test_comparison
], ignore_index=True)
metrics_table.to_csv(
    os.path.join(OUTPUT_DIR, "model_metrics_comparison.csv"),
    index=False
)
print("Complete comparison saved to:", os.path.join(OUTPUT_DIR, "model_metrics_comparison.csv"))

In [ ]:
# Save the selected model, preprocessing transformer, threshold, and feature list
selected_artifact = {
    "model": selected_model,
    "preprocessor": preprocessor,
    "feature_columns": feature_columns,
    "threshold": selected_threshold,
    "model_name": selected_name,
    "strategy": selected_strategy
}
selected_model_path = os.path.join(OUTPUT_DIR, "fraud_selected_model.joblib")
joblib.dump(selected_artifact, selected_model_path)
print("Selected model saved to:", selected_model_path)

In [ ]:
# Final comparison: baseline Logistic Regression vs. the selected candidate,
# whether there is a clear winner among candidates, and whether the selected
# candidate's test performance meets a business bar.
baseline_summary = {
    "model": "Logistic Regression",
    "strategy": "class_weight=balanced",
    "precision": test_precision,
    "recall": test_recall,
    "f1": test_f1,
    "f2": test_f2,
    "pr_auc": average_precision_score(y_test, test_probability),
    "roc_auc": roc_auc_score(y_test, test_probability),
}

comparison_metrics = ["precision", "recall", "f1", "f2", "pr_auc", "roc_auc"]
improvement = pd.DataFrame({
    "metric": comparison_metrics,
    "baseline_LR_test": [baseline_summary[m] for m in comparison_metrics],
    "selected_model_test": [test_metrics[m] for m in comparison_metrics],
})
improvement["absolute_gain"] = (
    improvement["selected_model_test"] - improvement["baseline_LR_test"]
)
print(f"Selected candidate: {selected_name} ({selected_strategy})")
display(improvement.round(4))



## Coverage Checklist



## Final Performance Comment

